readme.md
# Cross-platform C++ library for working with the X-Plane .obj format.

| CI        | master | develop | last commit |
| --------- |:------:|:-------:|:-----------:|
| appveyor  |[![Build status](https://ci.appveyor.com/api/projects/status/ym5rtb1ii6lhyric/branch/master?svg=true)](https://ci.appveyor.com/project/steptosky/xplnobj/branch/master)|[![Build status](https://ci.appveyor.com/api/projects/status/ym5rtb1ii6lhyric/branch/develop?svg=true)](https://ci.appveyor.com/project/steptosky/xplnobj/branch/develop)|[![Build status](https://ci.appveyor.com/api/projects/status/ym5rtb1ii6lhyric?svg=true)](https://ci.appveyor.com/project/steptosky/xplnobj)|
| travis    |[![Build Status](https://travis-ci.org/steptosky/XplnObj.svg?branch=master)](https://travis-ci.org/steptosky/XplnObj)|[![Build Status](https://travis-ci.org/steptosky/XplnObj.svg?branch=develop)](https://travis-ci.org/steptosky/XplnObj)|[![Build Status](https://travis-ci.org/steptosky/XplnObj.svg)](https://travis-ci.org/steptosky/XplnObj)|

---

- The library is distributed under [BSD (3-Clause)](http://opensource.org/licenses/BSD-3-Clause) license.
  For more information read the [license](license.txt) file.
- The library uses [SemVer](http://semver.org/).
- The library requires C++ 17 or higher.
- The library's versions from the ```master``` branch are available in our 
  [[bintray]](https://bintray.com/steptosky/conan-open-source) conan remote without pre-built packages.  
  ```XplnObj/X.Y.Z@steptosky/stable```
- The X-Plane [official website](http://www.x-plane.com/).
- The X-Plane [obj specification](http://developer.x-plane.com/?article=obj8-file-format-specification).

#### warning 
- As the library hasn't got the major version yet
  the library's interface and logic are not stable and can be changed at any time.


#### dependencies
- [Cmake](https://cmake.org) - building tool.
- [Conan](https://www.conan.io) - package manager.
- [Conan Package Tools](https://github.com/conan-io/conan-package-tools) - if you want to build the project with those tools.
- [Python 3](https://www.python.org) - is needed for the Conan.
- [Doxygen](http://www.stack.nl/~dimitri/doxygen) - if you want to generate the documentation.
- [Gtest and Gmock 1.8](https://github.com/google/googletest) - testing (is used via the conan) . 

#### documentation
Run from the root folder ```doxygen doxyfile``` the result will be in the ```doc-generated``` folder.  
The ```doxygen``` has to be accessible through your ```PATH``` environment variable.

#### environment variables
| Tools  | Variables | Type | Description |
|-------:|----------:|:----:|:------------|
| conan  | **CONAN_TESTING_REPORT_DIR** | _string_ | A path for storing tests results. Default value is specified in the cmake script. |
| conan  |      **CONAN_BUILD_TESTING** |  _0/1_   | Enables/disables building and running the tests.  If you set ```BUILD_TESTING=ON``` as a parameter while running ```cmake``` command it will auto-set ```CONAN_BUILD_TESTING=1```.  |
| cmake  |       **TESTING_REPORT_DIR** | _string_ | You can specify the directory for the tests reports, it can be useful for CI. Default value is specified in the cmake script. |
| cmake  |            **BUILD_TESTING** | _ON/OFF_ | Enables/disables building test projects. This is standard cmake variable. |

**Note:** sometimes you will need to delete the file ```cmake/conan.cmake``` then the newer version of this file will be downloaded from the Internet while running ```cmake``` command.  
This file is responsible for cmake and conan interaction.


## Building

- You have to install `conan` and [add](https://docs.conan.io/en/latest/reference/commands/misc/remote.html):  
    - If it isn't available [conan-center](https://bintray.com/conan/conan-center) remote the link can be `https://conan.bintray.com`.
    - [bincrafters bintray](https://bintray.com/bincrafters/public-conan) remote.

You may use some ways to build this library.
1. **Using cmake as usually.**   
    _These scripts are just examples probably you will need to adjust them for your purposes._  
    This approach will do:  
    - build the library or just generate project for your IDE.
    - test the library if you specify necessary environment variables.
    - install it into specified directory.
  
Usually this approach is for developing.  

**Examples:**  
Windows (.bat) For Visual Studio 2017.
``` batch
::==========================================================
@echo off
:: Fix problems with some symbols.
REM change CHCP to UTF-8
CHCP 1252
CLS
::==========================================================
:: Sometimes conan asks you about to login to private repositories
:: if you have added ones. So you can put into the root repository folder
:: the file called 'set-conan-user.bat' and write there something like this:
::      call conan user -p <password> -r <remote alias> <user name>
if exist set-conan-user.bat call set-conan-user.bat
::==========================================================
:: Creating building DIR and use it as working one.
set dir="msvc-2017"
if not exist %dir% mkdir %dir%
cd %dir%
::==========================================================
:: Generating Visual Studio project.
call cmake -G "Visual Studio 15 Win64" ../ ^
    -DBUILD_SHARED_LIBS=OFF ^
    -DCMAKE_INSTALL_PREFIX=../output ^
    -DBUILD_TESTING=ON
::==========================================================
:: Building
:: Keep it commented if you want to generate VS project only
:: (without building), otherwise uncomment it.
::call cmake --build . --target install --config Release
::call cmake --build . --target install --config Debug
::==========================================================
cd ../
pause
::==========================================================
```


Unix (Linux/Mac (.sh)).
``` bash
#===========================================================
# Setting building dir name
dir_name="build-release"
#===========================================================
# Creating building DIR and use it as working one.
rm -r ${dir_name}
mkdir ${dir_name}
cd ${dir_name}
#===========================================================
# Generating make files
cmake -G"Unix Makefiles" ../ \
    -DCMAKE_BUILD_TYPE="Release" \
    -DCMAKE_INSTALL_PREFIX=../output \
    -DBUILD_TESTING=ON
#===========================================================
# Building
cmake --build . --target install
#===========================================================
cd ../
#===========================================================
```
---

2. **Using [conan create](https://docs.conan.io/en/latest/reference/commands/creator/create.html).**  
    _These scripts are just examples probably you will need to adjust them for your purposes._   
    This approach will do:  
    - export conan recipe.
    - build the library in conan local cache.
    - test the library if you specify necessary environment variables.
    - test conan package (`test_package`).
    - as the result you have the library in your local conan cache in the channel you have specified.

Usually this approach is used when you want to build and put the library into conan local cache with certain settings.  
You may also [upload](https://docs.conan.io/en/latest/reference/commands/creator/upload.html) the built library into external conan remote later.

**Examples:**  
Windows (.bat) For Visual Studio 2017 x64.

``` batch
call conan create . steptosky/develop ^
     -s compiler="Visual Studio" ^
     -s compiler.version=15 ^
     -s compiler.runtime=MD ^
     -s build_type=Release ^
     -s arch=x86_64 ^
     -o XplnObj:shared=False ^
     -e CONAN_TESTING_REPORT_DIR="report/conan-test" ^
     -e CONAN_BUILD_TESTING=1 ^
     -e CTEST_OUTPUT_ON_FAILURE=1 ^
    --build=XplnObj

pause
:: remove build dir in test package
rd /s/q "test_package/build"
```

Unix (Linux/Mac (.sh)). Fix the example for you use case.
``` bash
conan create . steptosky/develop \
     -s compiler="apple-clang" \
     -s compiler.version=9.1 \
     -s build_type=Release \
     -s arch=x86_64 \
     -o XplnObj:shared=False \
     -e CONAN_TESTING_REPORT_DIR="report/conan-test" \
     -e CONAN_BUILD_TESTING=1 \
     -e CTEST_OUTPUT_ON_FAILURE=1 \
    --build=XplnObj
```
---


3. **You may also use just [conan export](https://docs.conan.io/en/latest/reference/commands/creator/export.html).**   
    This approach will do:  
    - export conan recipe to your local cache.
    - when project-consumer includes this library the necessary packaged will be auto-built.
    - as the result you have the library in your local conan cache in the channel you have specified.

Usually it may be used instead of previous issue when you don't want pre-build and test the library.

**Examples:**  
``` batch
conan . export steptosky/develop
```
---

4. **Using [conan package tools](https://github.com/conan-io/conan-package-tools).**   
    _These scripts are just examples probably you will need to adjust them for your purposes._   
    This approach will do:  
    - create matrix with various configurations.
    - run conan create for each configuration.
    - as the result you have the library in your local conan cache in the channel you have specified with various configurations.
    - the builds may also be auto-uploaded into external conan remote.

Usually this approach is used in CI for building various configurations.  

**Examples:**  
Windows (.bat) building with the conan package tools.
``` batch
::==========================================================
:: Sometimes conan asks you about to login to private repositories
:: if you have added ones. So you can put into the root repository folder
:: the file called 'set-conan-user.bat' and write there something like this:
::      call conan user -p <password> -r <remote alias> <user name>
if exist set-conan-user.bat call set-conan-user.bat
::==========================================================
set CONAN_USERNAME=steptosky
set CONAN_CHANNEL=testing
set CONAN_BUILD_TESTING=1
set CONAN_BUILD_POLICY=outdated
set CONAN_VISUAL_VERSIONS=15
set CONAN_ARCHS=x86_64
set CTEST_OUTPUT_ON_FAILURE=1
::==========================================================

call python build.py

:: uncomment next line if you want to remove all packages from all versions from specified channel
:: call conan remove XplnObj/*@steptosky/testing -f

pause

:: remove build dir in test package
rd /s/q "test_package/build"
```

Unix (Linux/Mac (.sh)). Fix the example for you use case.
``` bash
CONAN_USERNAME=steptosky
CONAN_CHANNEL=testing
CONAN_BUILD_TESTING=1
CONAN_BUILD_POLICY=outdated
CONAN_APPLE_CLANG_VERSIONS=9.1
CONAN_ARCHS=x86_64
CTEST_OUTPUT_ON_FAILURE=1

python build.py

# uncomment next line if you want to remove all packages from all versions from specified channel
# conan remove XplnObj/*@steptosky/testing -f
```
---

## Memo for the library developers
- [release-checklist](release-checklist.md) see this file when you are making the release.
- [change log](changelog.md) this file has to be filled during the release process and contains information about changes.
- [lasote docker hub](https://hub.docker.com/u/lasote/) - docker containers for some conan use cases, for example for [travis](.travis.yml) settings.
- Uploading to steptosky bintray:  
Windows (bintray-upload.bat) Example:
``` batch
ECHO "WARNING You are going to upload the conan recipe to bintray"
pause
::==========================================================
:: You have to put into the root repository folder
:: the file called 'set-conan-user.bat' and write there something like this:
::      conan user -p <api-key> -r sts-bintray <user>
if exist set-conan-user.bat call set-conan-user.bat
::==========================================================
conan export . steptosky/stable
conan upload XplnObj/[set-current-version]@steptosky/stable -r sts-bintray -c --retry 2 --retry-wait 5
::==========================================================
pause
  ```


## Copyright
Copyright (c) 2018, StepToSky team. All rights reserved.  
[www.steptosky.com](http://www.steptosky.com/)

changelog.md
StepToSky: X-Plane Obj Library
---------------------------------------------------------------------------
#### next

- **Added** Particle system support.
- **Added** Constraint for macos deployment target - 10.14, because of [error: call to unavailable function 'visit': introduced in macOS 10.14 std::visit].
- **Added** Checking if a normal map is specified but the attributes GLOBAL_specular or ATTR_shiny_rat aren't set.
- **Added** Checking if the ATTR_cockpit_device doesn't have a device name.
- **Added** Checking if mesh has the attribute ATTR_cockpit_region but the attribute COCKPIT_REGION isn't set.
- **Added** Helper for converting quaternion sequence to animation rotate.
- **Added** Parser for the [X-Plane Root/Resources/bitmaps/world/lites/lights.txt] file.
- **Fixed** Normals of mirrored objects.

##### Breaking backward compatibility:
- **Changed** Code style: public variable now have the prefix m instead of p.
- **Changed** Reworked animation classes.
- **Changed** Attribute sets use std::optional now.
- **Changed** 'Tree' attribute location. It is presented as a parameter of the mesh object now.
- **Changed** Manipulator use std::variant for its type now.
- **Removed** `enable/disable` flags from the attributes as the std::option is used.

---------------------------------------------------------------------------
#### 0.9.0-beta (27.11.2018)
##### Breaking backward compatibility:
- **Changed** Custom DataRefs and Commands format. The symbol ':' must immediate fallow ID.

---------------------------------------------------------------------------
#### 0.8.0-beta (25.10.2018)

- **Added** Classes for reading/writing datarefs and commands files.
- **Added** Context for importing/exporting.
- **Fixed** Fixed bug for attribute printing:  
            `AttrShiny`  
            `AttrBlend`  
            Default/disabled values were not printed when those attributes were enabled and then disabled.

##### Breaking backward compatibility:
- **Changed:** Methods for export/import in ObjMain.
- **Changed:** Under windows (MSVC) `std::wstring` is used as a path.
- **Removed:** Signature methods from the `ExportOptions`. Those methods are in the `ExportContext` now.

---------------------------------------------------------------------------
#### 0.7.0-beta (10.10.2018)

- **Added** Ability to use tool-tip in ATTR_manip_noop.
- **Added** Ability to use text variables in the param light's parameters string.
- **Added** ATTR_cockpit_device attribute.
- **Fixed** Sorting algorithm for LODs.

##### Breaking backward compatibility:
- **Changed:** Methods for accessing to the LODs.
- **Changed:** Changed named light's methods.
- **Changed:** Changed param light's methods.
- **Changed:** Changed Transform's methods.
- **Removed:** ELightNamed and ELightParams. 
               For getting light names use `X-Plane Root/Resources/bitmaps/world/lites/lights.txt`
 
---------------------------------------------------------------------------
#### 0.6.2-beta (31.08.2018)

- **Updated** Some manipulators UI text to be shorter.
- **Fixed** The library signature printing.
- **Fixed** Some items converting to string in `EManipulator`.

---------------------------------------------------------------------------
#### 0.6.1-beta (30.08.2018)

- **Changed** Enabling policy for `AttrAxisDetented` and `AttrManipWheel` manipulators.  

---------------------------------------------------------------------------
#### 0.6.0-beta (30.08.2018)

- **Added:** Ability to specifying strings that are printed before and after object.  
             This allows you to use new attributes manually until
             it is supported by the library or writing notes.
             This can be applied to any object types.
- **Added:** New manipulators:  
             `ATTR_manip_command_switch_left_right2`,  
             `ATTR_manip_command_switch_up_down2`,  
             `ATTR_manip_command_knob2`,  
             `ATTR_manip_drag_rotate`,  
             `ATTR_axis_detented`,  
             `ATTR_axis_detent_range`,  
             `ATTR_manip_keyframe`
- **Updated:** Some methods of manipulators were renamed/corrected according to .the obj specification. The old ones got the deprecated attribute.
- **Updated:** Building scripts and their documentation.
- **Fixed:** Printing incorrect attribute name for the
             `ATTR_manip_command_switch_up_down` and `ATTR_manip_command_switch_left_right`.
##### Breaking backward compatibility
- **Changes:** The wheel manipulator is composed now.  

---------------------------------------------------------------------------
#### 0.5.0-beta (05.10.2017)

- **Added:** Error if any LOD except the first one contains hard polygons.
- **Added:** Auto-flip normals for objects with mirrored transformation.
- **Fixed:** Creating two sided mesh.
- **Fixed:** Printing custom name of the `LIGHT_PARAM`.

---------------------------------------------------------------------------
#### 0.4.1-beta (24.08.2017)

- **Added:** Printing error if texture name contains illegal symbols.
- **Improved:** Build scripts.
- **Fixed:** Build scripts.

---------------------------------------------------------------------------
#### 0.4.0-beta (17.06.2017)

- **Added:** Smoke object printing. Although there was the smoke object before, 
             it could not be used. So it can be used now.
- **Added:** Export option for the smoke marking `XOBJ_EXP_MARK_SMOKE`.
- **Added:** Panel-Click manipulator. This manipulator should be used for 
             the cockpit geometry (`ATTR_cockpit`) when 
             you want to enable the panel manipulators.
- **Added:** Jenkins pipeline script.
- **Fixed:** Manipulators' state machine. There were some problems in the state machine 
             one of them could lead to undefined behavior it was fixed now, also 
             there was changed the default behavior. 
             By default all the manipulators are **disabled now**, 

conanfile.py
# ----------------------------------------------------------------------------------#
# //////////////////////////////////////////////////////////////////////////////////#
# ----------------------------------------------------------------------------------#
#
#  Copyright (C) 2018, StepToSky
#  All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are met:
#
#  1.Redistributions of source code must retain the above copyright notice, this
#    list of conditions and the following disclaimer.
#  2.Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and / or other materials provided with the distribution.
#  3.Neither the name of StepToSky nor the names of its
#    contributors may be used to endorse or promote products derived from
#    this software without specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
#  ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
#  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
#  DISCLAIMED.IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
#  ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
#  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
#  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
#  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
#  Contacts: www.steptosky.com
#
# ----------------------------------------------------------------------------------#
# //////////////////////////////////////////////////////////////////////////////////#
# ----------------------------------------------------------------------------------#

import os
from conans import ConanFile, CMake, tools
from vcs_info import VcsInfo


class LibConan(ConanFile):
    version = "0.9.0"
    name = 'XplnObj'
    url = 'https://github.com/steptosky/XplnObj'
    license = 'BSD 3-Clause'
    description = "This library is for working with the X-Plane's obj format. It requires C++ 17."
    author = 'StepToSky <info@steptosky.com>'

    settings = "os", "compiler", "build_type", "arch"

    options = {'shared': [True, False], "fPIC": [True, False]}
    default_options = 'shared=False', "fPIC=False", \
                      'stsff-logging:shared=False',\
                      'gtest:shared=False',\
                      'gtest:build_gmock=True'

    exports = 'vcs_data', 'vcs_info.py'
    exports_sources = 'CMakeLists.txt', 'src/*', 'src-test/*', 'src-test-resources/*', 'include/*', 'cmake/*', 'license*'

    generators = 'cmake'
    build_policy = 'missing'
        
    build_test_var = "CONAN_BUILD_TESTING"
    test_dir_var = "CONAN_TESTING_REPORT_DIR"
    vcs_data = VcsInfo()


    def configure(self):
        if self.settings.compiler == "Visual Studio" and float(str(self.settings.compiler.version)) < 15:
            raise Exception("Visual Studio 15 (2017) or higher is required")
        if self.settings.os != "Windows":
            self.options["stsff-logging"].fPIC = True

    def config_options(self):
        if self.settings.os == "Windows":
            del self.options.fPIC

    def requirements(self):
        self.requires("glm/0.9.9.4@g-truc/stable")
        self.requires("stsff-logging/1.0.2@steptosky/stable")
        if os.getenv(self.build_test_var, "0") == "1":
            self.requires('gtest/1.8.1@bincrafters/stable', private=True)

    def build(self):
        build_testing = os.getenv(self.build_test_var, "0")
        test_dir = os.getenv(self.test_dir_var, "")
        cmake = CMake(self)
        self.vcs_data.setup_cmake(cmake)
        cmake.definitions["BUILD_TESTING"] = 'ON' if build_testing == "1" else 'OFF'
        if test_dir:
            cmake.definitions["TESTING_REPORT_DIR"] = test_dir
        cmake.configure()
        cmake.build()
        cmake.install()
        if build_testing == "1":
            cmake.test(output_on_failure=True)

    def package(self):
        self.copy("license*", src=".", dst="licenses", ignore_case=True, keep_path=False)
        self.copy(pattern="*/%s.pdb" % self.name, dst='%s' % self.settings.build_type, src=".", keep_path=False)

    def package_info(self):
        lib_dir = '%s' % self.settings.build_type
        self.cpp_info.libdirs = [lib_dir]
        self.cpp_info.libs = tools.collect_libs(self, lib_dir)

# ----------------------------------------------------------------------------------#
# //////////////////////////////////////////////////////////////////////////////////#
# ----------------------------------------------------------------------------------#
